### DISCUSSION TITLE
freetube: AppImage does not work (missing --profile)

### DISCUSSION THREAD

**Discussion by ghost** [netblue30/firejail#5901] [Q&A] [answered]
I am posting this in discussion because I suspect I'm just doing something wrong and that it's not actually a problem with firejail. I mean it's a problem with firejail but because I'm using it wrong, I think at least.

Installed Manjaro XFCE minimal. It's the latest version.
Then installed Firejail from the official package manager.
Then ran firecfg. Tested with firefox and it's sandboxed and working.

Then downloaded the portable Freetube version 1.18. Then ran firecfg again but made no difference, makes sense because it's portable but did it anyway just in case.
I opened terminal from the Freetube directory. Ran freetube, it worked.
Then ran it again with firejail but got errors, don't remember what they were. Tried creating a symlink to the freetube file in the portable directory from /usr/bin/freetube. 

```sh
sudo ln -s ~/freetube-0.18.0-linux-portable-x64/freetube /usr/bin/freetube
```

Errors. What the errors were isn't important any longer because I moved on to appimage.

I downloaded the appimage version of freetube instead. I went to proporties and enabled it to be run as an executable.
I opened the directory in a terminal. Then ran it and it works. Then I ran it with `firejail --appimage` but then I get errors:

```
Reading profile /etc/firejail/et.profile
Reading profile /etc/firejail/wps.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Parent pid 9197, child pid 9202

** Warning: dropping all Linux capabilities and setting NO_NEW_PRIVS prctl **

Mounting appimage type 2
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Warning: /sbin directory link was not blacklisted
Warning: /usr/sbin directory link was not blacklisted
Warning: not remounting /run/user/1000/gvfs
Warning: Cannot confine the application using AppArmor.
Maybe firejail-default AppArmor profile is not loaded into the kernel.
As root, run "aa-enforce firejail-default" to load it.
Child process initialized in 540.69 ms
xdg-settings: default-url-scheme-handler not implemented for xfce
[3:0715/135510.961393:ERROR:process_singleton_posix.cc(481)] readlink failed: Permission denied (13)
[3:0715/135510.961715:ERROR:process_singleton_lock_posix.cc(20)] readlink(/home/lolol/.config/FreeTube/SingletonLock) failed: Permission denied (13)
[3:0715/135510.961826:ERROR:process_singleton_posix.cc(314)] readlink(/home/lolol/.config/FreeTube/SingletonLock) failed: Permission denied (13)
[3:0715/135510.961861:ERROR:process_singleton_posix.cc(338)] Failed to create /home/lolol/.config/FreeTube/SingletonLock: Permission denied (13)
[3:0715/135510.961940:ERROR:process_singleton_posix.cc(481)] readlink failed: Permission denied (13)
[3:0715/135510.961996:ERROR:process_singleton_lock_posix.cc(20)] readlink(/home/lolol/.config/FreeTube/SingletonLock) failed: Permission denied (13)
[3:0715/135511.193912:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/system: Permission denied
[3:0715/135511.194220:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/system: Permission denied
[3:0715/135511.232700:ERROR:address_tracker_linux.cc(189)] Could not create NETLINK socket: Operation not supported (95)
[3:0715/135511.233282:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
[3:0715/135511.233413:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/user: Permission denied
[3:0715/135511.287442:ERROR:udev_watcher.cc(52)] Failed to initialize a udev monitor.
(node:3) UnhandledPromiseRejectionWarning: Error: EACCES: permission denied, open '/home/lolol/.config/FreeTube/settings.db'
(Use `freetube --trace-warnings ...` to show where the warning was created)
(node:3) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:3) UnhandledPromiseRejectionWarning: Error: EACCES: permission denied, open '/home/lolol/.config/FreeTube/profiles.db'
(node:3) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 4)
(node:3) UnhandledPromiseRejectionWarning: Error: EACCES: permission denied, open '/home/lolol/.config/FreeTube/history.db'
(node:3) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 6)
(node:3) UnhandledPromiseRejectionWarning: Error: EACCES: permission denied, open '/home/lolol/.config/FreeTube/playlists.db'
(node:3) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 8)
[Error: EACCES: permission denied, open '/home/lolol/.config/FreeTube/settings.db'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '/home/lolol/.config/FreeTube/settings.db'
}

Parent is shutting down, bye...
AppImage detached
```

So it's strange it doesn't read the freetube profile in there. When I tried using the portable version it would at least read the freetube profile.

So I'm guessing that at least a part of the problem is that freetube profile isn't being read. But I suspect there are more problems. 

I want help with getting the appimage version working. Don't care about the portable version any more.

I also would like to have a desktop entry for the appimage. But first I want to get it working without a desktop entry file. But when it's time I have prepared this:

```ini
[Desktop Entry]
Name=freetube
Exec=firejail --appimage ~/applications/freetube_0.18.0_amd64.AppImage %U
Terminal=false
Type=Application
Categories=Network;Application;
```

I don't think the `%U` thing is needed actually but I guess it doesn't harm. Not exactly sure what it does but just leaving it as it is for now.

Also in the errors/warnings there's something about apparmor as well. I never have enabled it. I'm doing all this on a VM. So this is just for testing/learning. After I have figured it out I will do it on my host computer. There is another app that doesn't have any security profile so that's another reason for doing it in a testing VM so I can create the profile there and later copy it to the host before installing the program there as well. But point is that I am hoping I don't need to enable apparmor. It's a bit complicated process I think with editing some of the files and restarting like 2 times or something as well. I will do it on the host though.

I actually tried to enable apparmor like it suggests in the error/warning messages by typing: sudo aa-enforce firejail-default
But it gave error:

```
Setting /etc/apparmor.d/firejail-default to enforce mode.
ERROR: Cache read/write disabled: interface file missing. (Kernel needs AppArmor 2.4 compatibility patch.)
Warning: unable to find a suitable fs in /proc/mounts, is it mounted?
Use --subdomainfs to override.
```

So I guess I would have to do it the longer complicated way but I hope it's not necessary for the VM where I just want to get firejail working. As I understand, you are supposed to be able to use only firejail without apparmor. That apparmor is optional.

**Comment by rusty-snake** [comment, marked as answer]
AppArmor is optional, correct. If firejail is compiled with AppArmor support but it is not working it warns you because it could be a configuration issue that you want to fix. If you do not want/use AA there's nothing you have to do.

Regarding the profile, if you use AppImages you have to specify the profile via `--profile=freetube`. We have a heuristic. But heuristics can fail.

**Reply by ghost** [reply]
Thanks, that worked. Still learning about Firejail. I'm getting better though. Finally starting to understand symbolic links and which directories precede others etc. I'm probably exactly the kind of user you are building firejail for. Not a sysadmin but still somewhat tech savvy, basically just a user. 

I also changed the Exec line in the desktop entry so now that works as well:

```ini
Exec=firejail --appimage --profile=freetube /home/lolol/applications/freetube_0.18.0_amd64.AppImage %U
```

Still get a lot of warnings and errors even though freetube is running. Like /sbin and /usr/sbin not being blacklisted sounds bad. Are all these warnings and errors because I am not using AppArmor?

The way I understood is that firejail when working together with apparmor is kind of duplicating some of the apparmor things, so if apparmor has a rule preventing X, but apparmor would fail for some reason, then firejail is also doing the same thing and preventing X. Or have I misunderstood how it works? Maybe firejail skips doing some things because it expects apparmor to do them?

This is the output I get now with Freetube running:

```
Reading profile /etc/firejail/freetube.profile
Reading profile /etc/firejail/allow-bin-sh.inc
Reading profile /etc/firejail/electron.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/disable-xdg.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Parent pid 10136, child pid 10141

** Warning: dropping all Linux capabilities and setting NO_NEW_PRIVS prctl **

Mounting appimage type 2
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Warning: skipping alternatives for private /etc
Warning: skipping asound.conf for private /etc
Warning: skipping crypto-policies for private /etc
Warning: skipping gtk-2.0 for private /etc
Warning: skipping ld.so.preload for private /etc
Warning: skipping pki for private /etc
Private /etc installed in 143.14 ms
Private /usr/etc installed in 0.01 ms
Warning: /sbin directory link was not blacklisted
Warning: /usr/sbin directory link was not blacklisted
Warning: Cannot confine the application using AppArmor.
Maybe firejail-default AppArmor profile is not loaded into the kernel.
As root, run "aa-enforce firejail-default" to load it.
Child process initialized in 605.39 ms
xdg-settings: default-url-scheme-handler not implemented for xfce
[20:0715/150306.366919:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/system: Permission denied
[20:0715/150306.367114:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/system: Permission denied
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
[55:0715/150306.827629:ERROR:viz_main_impl.cc(186)] Exiting GPU process due to errors during initialization
TypeError: Cannot destructure property 'value' of 'object null' as it is null.
    at /run/firejail/appimage/resources/app.asar/dist/main.js:2:240841
    at async _ (/run/firejail/appimage/resources/app.asar/dist/main.js:2:240797)
    at async App.<anonymous> (/run/firejail/appimage/resources/app.asar/dist/main.js:2:247098)
(node:20) UnhandledPromiseRejectionWarning: Error: Cannot create a secure cookie from an insecure URL
(Use `freetube --trace-warnings ...` to show where the warning was created)
(node:20) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:20) UnhandledPromiseRejectionWarning: Error: Cannot create a secure cookie from an insecure URL
(node:20) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
[84:0715/150307.433069:ERROR:viz_main_impl.cc(186)] Exiting GPU process due to errors during initialization
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
[105:0715/150307.648791:ERROR:gpu_memory_buffer_support_x11.cc(44)] dri3 extension not supported.

(freetube:20): IBUS-WARNING **: 15:03:18.368: Unable to load /var/lib/dbus/machine-id: Failed to open file “/var/lib/dbus/machine-id”: No such file or directory
```

**Reply by rusty-snake** [reply]
>  Are all these warnings and errors because I am not using AppArmor?

No. Only the three lines talk about AA are because of AA.

> The way I understood is that firejail when working together with apparmor is kind of duplicating some of the apparmor things, so if apparmor has a rule preventing X, but apparmor would fail for some reason, then firejail is also doing the same thing and preventing X. Or have I misunderstood how it works? Maybe firejail skips doing some things because it expects apparmor to do them?

Yes and Yes and No.

There are two things.
1. Use firejails AA integration which runs the sandbox as usually but in `firejail-default` (AppArmor profile) for additional hardening.
2. Use firejail and AA with a own profile (say `usr.bin.firefox`). disadviced.

> ```
> ** Warning: dropping all Linux capabilities and setting NO_NEW_PRIVS prctl **
> ```

Info from firejail because you use `--appimage`.

> ```
> Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
> ```

Warning from firejail about a sandbox hole. 99.9% of use see this.

> ```
> Warning: skipping ... for private /etc
> ```

Info from firejail. Can be ignored.

> ```
> Warning: /sbin directory link was not blacklisted
> Warning: /usr/sbin directory link was not blacklisted
> ```

Not sure why. But you can blacklist them in freetube.local if you want.

> ```
> Warning: Cannot confine the application using AppArmor.
> Maybe firejail-default AppArmor profile is not loaded into the kernel.
> As root, run "aa-enforce firejail-default" to load it.
> ```

Warning/Info from firejail. AA.

> ```
> xdg-settings: default-url-scheme-handler not implemented for xfce
> [20:0715/150306.366919:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/system: Permission denied
> [20:0715/150306.367114:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/firejail/mnt/dbus/system: Permission denied
> libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
> [55:0715/150306.827629:ERROR:viz_main_impl.cc(186)] Exiting GPU process due to errors during initialization
> TypeError: Cannot destructure property 'value' of 'object null' as it is null.
> at /run/firejail/appimage/resources/app.asar/dist/main.js:2:240841
> at async _ (/run/firejail/appimage/resources/app.asar/dist/main.js:2:240797)
> at async App. (/run/firejail/appimage/resources/app.asar/dist/main.js:2:247098)
> (node:20) UnhandledPromiseRejectionWarning: Error: Cannot create a secure cookie from an insecure URL
> (Use freetube --trace-warnings ... to show where the warning was created)
> (node:20) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag --unhandled-rejections=strict (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
> (node:20) UnhandledPromiseRejectionWarning: Error: Cannot create a secure cookie from an insecure URL
> (node:20) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag --unhandled-rejections=strict (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
> libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
> [84:0715/150307.433069:ERROR:viz_main_impl.cc(186)] Exiting GPU process due to errors during initialization
> libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
> [105:0715/150307.648791:ERROR:gpu_memory_buffer_support_x11.cc(44)] dri3 extension not supported.
> (freetube:20): IBUS-WARNING **: 15:03:18.368: Unable to load /var/lib/dbus/machine-id: Failed to open file “/var/lib/dbus/machine-id”: No such file or directory
> ```

From freetube. Some are because of firejail, some not. As long as everything works they can be ignored.
